api/clients/v2/validator/retrieval_worker.go

```diff
@@ -0,0 +1,606 @@
+func (w *retrievalWorker) handleCompletedDownload(message *downloadCompleted) {
+	if message.Err == nil {
+		if w.config.DetailedLogging {
+			w.logger.Debug("downloaded chunks from operator",
+				"operator", message.OperatorID.Hex(),
+				"blobKey", w.blobKey.Hex())
+		}
+		w.downloadedChunksQueue.Enqueue(message)
+		w.updateChunkStatus(message.OperatorID, downloaded)
+	} else {
+		w.logger.Warn("failed to download chunk data",
+			"operator", message.OperatorID.Hex(),
+			"blobKey", w.blobKey.Hex(),
+			"err", message.Err)
+
+		w.updateChunkStatus(message.OperatorID, failed)
+	}
+}
+
+// handleVerificationCompleted handles the completion of a verification.
+func (w *retrievalWorker) handleVerificationCompleted(message *verificationCompleted) {
+	if message.Err == nil {
+		if w.config.DetailedLogging {
+			w.logger.Debug("verified chunks from operator",
+				"operator", message.OperatorID.Hex(),
+				"blobKey", w.blobKey.Hex())
+		}
+		w.verifiedChunksQueue.Enqueue(message)
+		w.updateChunkStatus(message.OperatorID, verified)
+	} else {
+		w.logger.Warn("failed to verify chunk data",
+			"operator", message.OperatorID.Hex(),
+			"blobKey", w.blobKey.Hex(),
+			"err", message.Err)
+
+		w.updateChunkStatus(message.OperatorID, failed)
+	}
+}
+
+// scheduleDownloads schedules downloads as needed.
+func (w *retrievalWorker) scheduleDownloads() {
+	for w.nextDownloadIndex < len(w.downloadOrder) {
+		if w.getStatusCount(downloading, downloaded, verifying, verified) >= w.targetDownloadCount {
+			// We've requested enough downloads
+			break
+		}
+		operatorID := w.downloadOrder[w.nextDownloadIndex]
+		w.updateChunkStatus(operatorID, downloading)
+
+		w.connectionPool.Submit(func() {
+			w.downloadChunks(operatorID)
+		})
+
+		w.nextDownloadIndex++
+	}
+}
+
+// scheduleVerifications schedules verifications as needed.
+func (w *retrievalWorker) scheduleVerifications() {
+	for !w.downloadedChunksQueue.Empty() {
+		if w.getStatusCount(verifying, verified) >= w.targetVerifiedCount {
+			// We've requested enough verifications
+			break
+		}
+
+		next, _ := w.downloadedChunksQueue.Dequeue()
+		reply := next.(*downloadCompleted).Reply
+		operatorID := next.(*downloadCompleted).OperatorID
+
+		w.updateChunkStatus(operatorID, verifying)
+
+		w.computePool.Submit(func() {
+			w.deserializeAndVerifyChunks(w.ctx, operatorID, reply)
+		})
+	}
+}
+
+// decodeBlob decodes the blob from the chunks.
+func (w *retrievalWorker) decodeChunks() ([]byte, error) {
+	w.probe.SetStage("decode")
+
+	if w.config.DetailedLogging {
+		w.logger.Info("decoding blob", "blobKey", w.blobKey.Hex())
+	}
+
+	chunks := make([]*encoding.Frame, 0)
+	indices := make([]uint, 0)
+
+	for !w.verifiedChunksQueue.Empty() {
+		next, _ := w.verifiedChunksQueue.Dequeue()
+		operatorID := next.(*verificationCompleted).OperatorID
+		operatorChunks := next.(*verificationCompleted).chunks
+
+		assignment := w.assignments[operatorID]
+		assignmentIndices := make([]uint, len(assignment.GetIndices()))
```

Comment:
You're calling `assignment.GetIndices()` twice, once to get the length, and once to iterate over the values. But `assignment.GetIndices()` appears to have a non-trivial cost, probably worth creating a local variable

Reply:
fixed